add from_ckpt method as Mixin

1lint opened the pull request:
Added a `from_ckpt` method to the stable diffusion pipeline classes that invokes the convert_from_ckpt logic to load pipeline weights from the original stable diffusion checkpoint format. Pipelines can inherit the method through the FromCkptMixin class. 

Also added logic to load checkpoint from HF hub, small edits to infer safetensor format, support pruned checkpoint files. 

Example:
```py
from diffusers import StableDiffusionPipeline
import torch

checkpoint_link = "https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/v1-5-pruned-emaonly.ckpt"
sd_pipe = StableDiffusionPipeline.from_ckpt(checkpoint_link)
sd_pipe.to("cuda")

output = sd_pipe('an astronaut riding horse')
output.images[0]
```

I found that loading checkpoint directly to cuda (default behavior of convert_from_ckpt.py) and then calling pipe.to('cuda') causes duplicative weights in VRAM, which crashed my system. I changed the logic to load checkpoint weights to cpu and then set the device using the pipeline.to() method.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2318). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented:
@williamberman could you take a look here?

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -828,13 +860,50 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
     if prediction_type == "v-prediction":
         prediction_type = "v_prediction"
 
+    if from_safetensors is None:
+        from_safetensors = checkpoint_path.rsplit(".", 1)[-1] == "safetensors"
```
I might be missing it but where do we use `from_safetensors`?

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -17,32 +17,57 @@
+
 from ...utils.import_utils import BACKENDS_MAPPING
 
+# silences transformers warning on CLIP Model loading unused keys
+logging.set_verbosity_error()
```
```suggestion
```

Silencing errors should be job of script that calls this script. Agreed the CLIP unused key errors are annoying :)

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -644,6 +669,10 @@ def convert_ldm_clip_checkpoint(checkpoint):
         if key.startswith("cond_stage_model.transformer"):
             text_model_dict[key[len("cond_stage_model.transformer.") :]] = checkpoint[key]
 
+    # if missing text encoder keys, checkpoint may be pruned so just use pretrained weights
+    if text_model_dict:
+        text_model.load_state_dict(text_model_dict)
+
     text_model.load_state_dict(text_model_dict)
```
I'm not sure I follow, it looks like this is calling load_state_dict twice regardless?

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -871,7 +938,7 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
             key_name = "model.diffusion_model.input_blocks.2.1.transformer_blocks.0.attn2.to_k.weight"
 
-            original_config_file = os.path.join(tmpdir, "inferenc.yaml")
+            original_config_file = os.path.join(tmpdir, "inference.yaml")
```
If you rebase on main, this should be taken car of :)

1lint commented:
Sorry for delayed comments, I typed them yesterday and they showed up in line on my view but they were "pending" (I thought they were publicly visible, still getting used to pull request interface 😅).

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1001,21 +1067,38 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
-    return pipe
+    if device is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+    return pipe.to(device)
```
```suggestion
    return pipe
```
The user should have full control here :-) Let's not move the pipeline to GPU if it exists

1lint replied:
sounds good! I will also run a rebase on the controlnet changes

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1001,21 +1067,38 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
         pipe = LDMTextToImagePipeline(vqvae=vae, bert=text_model, tokenizer=tokenizer, unet=unet, scheduler=scheduler)
 
-    return pipe
+    if device is None:
```
```suggestion
```

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1001,21 +1067,38 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
 
-    return pipe
+    if device is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
```
```suggestion
```

patrickvonplaten reviewed (commented):
This looks nice already! Could you also maybe add  a test that showcases how the downloaded works? :-) 

We could add the test here: https://github.com/huggingface/diffusers/blob/664b4de9e22a825875b6bad45867f8f77cdf95d6/tests/test_pipelines.py#L62

Maybe you can use two checkpoints as examples:
- https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/v1-5-pruned.ckpt
- https://huggingface.co/WarriorMama777/OrangeMixs/blob/main/Models/AbyssOrangeMix/AbyssOrangeMix_base.ckpt

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -45,7 +46,7 @@
 
 
-class StableDiffusionPipeline(DiffusionPipeline):
+class StableDiffusionPipeline(DiffusionPipeline, FromCkptMixin):
```
very nice!

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -828,13 +856,50 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
+        use_auth_token = kwargs.pop("use_auth_token", None)
+        revision = kwargs.pop("revision", None)
+
+        cached_folder = snapshot_download(
```
actually we should use `snapshot_download` here since we're just downloading a single .ckpt file. Could you instead make use of: https://huggingface.co/docs/huggingface_hub/v0.12.1/en/package_reference/file_download#huggingface_hub.hf_hub_download

More specifically the following should work:

```py
file = hf_hub_download(
            repo_id,
            filename=file_path,
            cache_dir=cache_dir,
            resume_download=resume_download,
            proxies=proxies,
            local_files_only=local_files_only,
            use_auth_token=use_auth_token,
            revision=revision,
        )
```

1lint replied:
Yes will do, thanks for the link!

patrickvonplaten reviewed (commented):
Let's try to use `hf_hub_download` instead of `snapshot_download`

Super nice work though :heart_eyes:

1lint commented:
Appreciate you keeping this PR in mind, I pushed the updates into this branch after rebasing on main 😄 

I added the FromCkpt Mixin to just the few stable diffusion pipeline classes that I tested, but I believe it should work with some other pipeline classes in the same folder (I don't have checkpoint files to test all the classes). Some classes may need minor tweaks or supporting logic in convert_from_ckpt.py (i.e. StableDiffusionLatentUpscalePipeline does not accept `safety_checker` but otherwise uses the same checkpoint weights as StableDiffusionPipeline). 

I added a download test for the suggested checkpoint files. In my own testing I used the following for quick testing with smaller files.

```
from diffusers import StableDiffusionPipeline
import tempfile
import os

with tempfile.TemporaryDirectory() as tmpdirname:

    ckpt_paths = [
        "hf-internal-testing/tiny-stable-diffusion-torch/vae/config.json",
        "hf-internal-testing/stable-diffusion-all-variants/model_index.json",
    ]

    for ckpt_path in ckpt_paths:
        try:
            _ = StableDiffusionPipeline.from_ckpt(
                ckpt_path, cache_dir=tmpdirname
            )
        except:
            pass

    ckpt_names = [os.path.basename(ckpt_path) for ckpt_path in ckpt_paths]

    files = []
    for cache in os.listdir(tmpdirname):
        snapshots = os.path.join(tmpdirname, cache, "snapshots")
        all_root_files = [t[-1] for t in os.walk(snapshots)]
        files += [item for sublist in all_root_files for item in sublist]

    # check that downloaded filenames match checkpoint filenames
    assert set(ckpt_names) == set(files)
```

1lint commented:
Most recent commit adds the FromCkpt Mixin to the controlnet pipeline.

I also added logic to infer and download the controlnet configs, and think it makes sense here to use HF Hub for storing/hosting the stable diffusion configs. This way the user doesn't need to download the configs from Github each time they use `load_pipeline_from_original_stable_diffusion_ckpt`. 
I think it would be great to host the configs on an official account. Right now I have them saved at https://huggingface.co/lint/stable_diffusion_configs

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1287,3 +1373,14 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
+    def from_ckpt(cls, model_path_or_checkpoint, **kwargs):
+        # silences transformers warning on CLIP Model loading unused keys
+
+        logging.set_verbosity_error()
```
let's not silence logs in a class method like this. That should be the job of the user if they want to. Especially as `logging.set_verbosity_error` is mutative and will permanently change the logging verbosity for the rest of the process

1lint replied:
Okay, makes sense! I've removed the logging setting in the latest commit

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1046,30 +1134,28 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
+                config_file_name = "cldm_v15.yaml"
 
+        original_config_file = hf_hub_download(
+            'lint/stable_diffusion_configs',
```
we shouldn't use a personal hub repo

1lint replied:
Got it, I've added a comment note to change the repo link in the final commit. If its a good idea to host it on hub, lmk what repo link to put in

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1007,13 +1016,92 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
+    if from_safetensors is None:
+        from_safetensors = checkpoint_path.rsplit(".", 1)[-1] == "safetensors"
+
+    # kwargs used for downloading both ckpt and config files
+    proxies = kwargs.pop("proxies", None)
+    use_auth_token = kwargs.pop("use_auth_token", None)
+
+    # Code based on diffusers.pipelines.pipeline_utils.DiffusionPipeline.from_pretrained
+    if not os.path.isfile(checkpoint_path):
+        checkpoint_path = Path(checkpoint_path)
+
+        # get repo_id and (potentially nested) file path of ckpt in repo
+        repo_id = str(Path().joinpath(*checkpoint_path.parts[:2]))
+        file_path = str(Path().joinpath(*checkpoint_path.parts[2:]))
+
+        cache_dir = kwargs.pop("cache_dir", DIFFUSERS_CACHE)
+        resume_download = kwargs.pop("resume_download", False)
+        local_files_only = kwargs.pop("local_files_only", HF_HUB_OFFLINE)
+        revision = kwargs.pop("revision", None)
+
+        checkpoint_path = hf_hub_download(
+            repo_id,
+            filename=file_path,
+            cache_dir=cache_dir,
+            resume_download=resume_download,
+            proxies=proxies,
+            local_files_only=local_files_only,
+            use_auth_token=use_auth_token,
+            revision=revision,
+        )
```
we shouldn't need to make this many changes to the existing function. any changes for downloading from the hub can be self contained in the calling function

williamberman commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1007,13 +1016,92 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
+    from diffusers import (
+        StableDiffusionControlNetPipeline,
+        StableDiffusionPipeline,
+        StableUnCLIPImg2ImgPipeline,
+        StableUnCLIPPipeline,
+        LDMTextToImagePipeline,
+        PaintByExamplePipeline
+    )
+
+    if pipeline_class:
+
+        pipeline_name = pipeline_class.__name__
+
+        if pipeline_name == "StableDiffusionControlNetPipeline":
+            model_type = "FrozenCLIPEmbedder"
+            controlnet = True
+
+        elif "StableDiffusion" in pipeline_name:
+            model_type = "FrozenCLIPEmbedder"
+            controlnet = False
+
+        elif pipeline_name == "StableUnCLIPImg2ImgPipeline":
+            model_type == "FrozenOpenCLIPEmbedder"
+            stable_unclip = "img2img"
+
+        elif pipeline_name == "StableUnCLIPPipeline":
+            model_type == "FrozenOpenCLIPEmbedder"
+            stable_unclip = "txt2img"
+
+        elif pipeline_name == "PaintByExamplePipeline":
+            model_type == "PaintByExample"
+
+        elif pipeline_name == "LDMTextToImagePipeline":
+            model_type == "LDMTextToImage"
+
+        else:
+            raise ValueError(f"Unhandled pipeline class: {pipeline_name}")
+
+    else:
+        pipeline_class = StableDiffusionPipeline
```
I'm not really a fan of this. If the model_type and other parameters have to be set, they should just be passed in as arguments and not overridden by the pipeline.

If the goal is just to support pipeline classes other than the default ones used by the script, just pass in the class and use it in lieu of the other constructor.

i.e.
```py
if pipeline_class is not None:
    pipeline_class(<args>)
else:
   StableDiffusionPipeline(<args>) # <- or other pipeline that is used by default
```

1lint replied:
Yes good point, it is duplicative with the `model_type` parameter. I've moved this logic down to FromCkptMixin as well

patrickvonplaten commented:
Nice progress @1lint, please let me know if you need any help on any of @williamberman suggestions :-)

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py:
```diff
@@ -1007,13 +1016,35 @@ def load_pipeline_from_original_stable_diffusion_ckpt(
     """
+
+    # import pipelines here to avoid circular import error when using from_ckpt method
+    from diffusers import (
```
ok for me!

patrickvonplaten reviewed (approved):
Very nice - this looks good to me! @williamberman can you double check as well?

patrickvonplaten commented:
Hey @1lint,

Very sorry for the delay here, I'll prio this PR now and try to help you get it merged.

patrickvonplaten commented on docs/source/en/api/pipelines/stable_diffusion/text2img.mdx:
```diff
@@ -39,6 +39,8 @@ Available Checkpoints are:
 	- enable_vae_tiling
 	- disable_vae_tiling
+  - from_ckpt
+  - load_attn_procs
```
```suggestion
```

alexblattner commented:
I tried to use from_ckpt on the controlnet pipeline and it doesn't seem to recognize it AttributeError: type object 'ControlNetModel' has no attribute 'from_ckpt'

1lint commented:
Congrats/thanks for merging! I can see a lot of additional polish/refactoring went into integrating the feature.

@alexblattner 
You could convert the controlnet checkpoint with the following
```
from diffusers.pipelines.stable_diffusion.convert_from_ckpt import download_controlnet_from_original_ckpt
from urllib.request import urlretrieve

controlnet_ckpt_path = "path/to/ckpt"
config_file = "cldm_v15.yaml" # set to "cldm_v21.yaml" if you are using stable diffusion V2

urlretrieve(f"https://raw.githubusercontent.com/lllyasviel/ControlNet/main/models/{config_file}", config_file)
pipe = download_controlnet_from_original_ckpt(controlnet_ckpt_path, original_config_file=config_file)
```

The original controlnet models use a separate config file so it can't support the from_ckpt method (without additional changes to the conversion code)

alexblattner commented:
@1lint I don't have the yaml only the safetensor. I am literally using a checkpoint from civitai

alexblattner commented:
@1lint I am using this code:

```
controlnet = ControlNetModel.from_pretrained("lllyasviel/control_v11p_sd15_openpose", torch_dtype=torch.float16)
pipe = StableDiffusionControlNetPipeline.from_pretrained(
  "local safetensor",controlnet=controlnet, torch_dtype=torch.float16,safety_checker=None, requires_safety_checker=False,
).to("cuda")
```
I don't think your example helps in this case.
